File: Sources/Services/FactorLifecycleService.swift
import Foundation

final class FactorLifecycleService {
    func evaluate(
        currentState: FactorState,
        evidence: FactorLifecycleEvidence
    ) -> FactorLifecycleDecision {
        if evidence.dataContamination || evidence.lookAheadBias ||
            evidence.trainTestLeakage ||
            evidence.unreproducibleResult ||
            evidence.definitionError || !evidence.trialRecordPresent {
            return decision(
                .quarantined,
                global: true,
                blocks: true,
                multiplier: 0,
                reason: "Immediate global quarantine: temporal, definition, data, or trial-record integrity failed."
            )
        }
        if currentState == .quarantined {
            return decision(
                .quarantined,
                global: true,
                blocks: true,
                multiplier: 0,
                reason: "Quarantine is global and requires a new reviewed definition."
            )
        }
        if currentState == .retired {
            if evidence.renewedEvidence {
                return decision(
                    .shadow,
                    global: false,
                    blocks: true,
                    multiplier: 0,
                    reason: "Renewed evidence permits Shadow observation, never a direct jump to Active."
                )
            }
            return decision(
                .retired,
                global: false,
                blocks: true,
                multiplier: 0,
                reason: "Retired factor remains blocked without renewed evidence."
            )
        }
        if evidence.changePointProbability >= 0.80 &&
            evidence.counterfactualImprovement >= 0.02 {
            return decision(
                .retired,
                global: false,
                blocks: true,
                multiplier: 0,
                reason: "Retired after change-point and counterfactual replacement evidence."
            )
        }
        if evidence.repeatedOosFailures >= 2 {
            return decision(
                .probation,
                global: false,
                blocks: true,
                multiplier: 0.15,
                reason: "Repeated purged OOS deterioration triggered Probation."
            )
        }
        if evidence.regimeSpecificFailure {
            return decision(
                currentState,
                global: false,
                blocks: true,
                multiplier: min(0.50, riskMultiplier(currentState)),
                reason: "Regime Conditional evidence recorded; global retirement was not triggered."
            )
        }
        if evidence.shortTermScore < 0.35 &&
            evidence.longTermScore >= 0.60 {
            return decision(
                .reduced,
                global: false,
                blocks: true,
                multiplier: 0.50,
                reason: "Short-term deterioration with stable long-term evidence triggered automatic risk contraction."
            )
        }
        if currentState == .reduced || currentState == .probation {
            if evidence.renewedEvidence &&
                evidence.shortTermScore >= 0.60 &&
                evidence.longTermScore >= 0.65 {
                return decision(
                    .shadow,
                    global: false,
                    blocks: true,
                    multiplier: 0.25,
                    reason: "Renewed evidence permits Shadow review; risk expansion still requires promotion evidence."
                )
            }
            return decision(
                currentState,
                global: false,
                blocks: true,
                multiplier: riskMultiplier(currentState),
                reason: "Risk expansion remains blocked pending renewed evidence."
            )
        }
        return decision(
            currentState,
            global: false,
            blocks: currentState != .active,
            multiplier: riskMultiplier(currentState),
            reason: "Evidence is stable; no lifecycle contraction is required."
        )
    }

    private func decision(
        _ state: FactorState,
        global: Bool,
        blocks: Bool,
        multiplier: Double,
        reason: String
    ) -> FactorLifecycleDecision {
        FactorLifecycleDecision(
            state: state,
            global: global,
            blocksRiskExpansion: blocks,
            riskMultiplier: multiplier,
            reason: reason
        )
    }

    private func riskMultiplier(_ state: FactorState) -> Double {
        switch state {
        case .active: return 1
        case .reduced: return 0.50
        case .probation: return 0.15
        case .candidate, .shadow, .retired, .quarantined: return 0
        }
    }
}

File: Windows/Services/FactorLifecycleService.cs
namespace CytisusTrading.Windows;

public interface IFactorLifecycleService
{
    FactorLifecycleDecision Evaluate(
        FactorState currentState,
        FactorLifecycleEvidence evidence);
}

public sealed class FactorLifecycleService : IFactorLifecycleService
{
    public FactorLifecycleDecision Evaluate(
        FactorState currentState,
        FactorLifecycleEvidence evidence)
    {
        if (evidence.DataContamination ||
            evidence.LookAheadBias ||
            evidence.TrainTestLeakage ||
            evidence.UnreproducibleResult ||
            evidence.DefinitionError ||
            !evidence.TrialRecordPresent)
        {
            return new FactorLifecycleDecision(
                FactorState.Quarantined,
                true,
                true,
                0,
                evidence.DataContamination
                    ? "Global quarantine: data contamination."
                    : evidence.LookAheadBias
                        ? "Global quarantine: look-ahead bias."
                        : evidence.TrainTestLeakage
                            ? "Global quarantine: train-test leakage."
                            : evidence.UnreproducibleResult
                                ? "Global quarantine: unreproducible result."
                                : evidence.DefinitionError
                                    ? "Global quarantine: definition error."
                                    : "Global quarantine: missing trial record.");
        }

        if (currentState == FactorState.Quarantined)
        {
            return new FactorLifecycleDecision(
                FactorState.Quarantined,
                true,
                true,
                0,
                "Quarantine is global and requires a new reviewed definition.");
        }

        if (currentState == FactorState.Retired)
        {
            return evidence.RenewedEvidence
                ? new FactorLifecycleDecision(
                    FactorState.Shadow,
                    false,
                    true,
                    0,
                    "Renewed evidence permits Shadow observation, never a direct jump to Active.")
                : new FactorLifecycleDecision(
                    FactorState.Retired,
                    false,
                    true,
                    0,
                    "Retired factor remains blocked without renewed evidence.");
        }

        if (evidence.ChangePointProbability >= 0.80 &&
            evidence.CounterfactualImprovement >= 0.02)
        {
            return new FactorLifecycleDecision(
                FactorState.Retired,
                false,
                true,
                0,
                "Retired after a high change-point probability and sustained counterfactual improvement after removal.");
        }

        if (evidence.RepeatedOosFailures >= 2)
        {
            return new FactorLifecycleDecision(
                FactorState.Probation,
                false,
                true,
                0.15,
                "Repeated purged OOS deterioration triggered Probation.");
        }

        if (evidence.RegimeSpecificFailure)
        {
            return new FactorLifecycleDecision(
                currentState,
                false,
                true,
                Math.Min(0.50, RiskMultiplier(currentState)),
                "Regime Conditional evidence recorded; global retirement was not triggered.");
        }

        if (evidence.ShortTermScore < 0.35 &&
            evidence.LongTermScore >= 0.60)
        {
            return new FactorLifecycleDecision(
                FactorState.Reduced,
                false,
                true,
                0.50,
                "Short-term deterioration with stable long-term evidence triggered automatic risk contraction.");
        }

        if (currentState is FactorState.Reduced or FactorState.Probation)
        {
            if (evidence.RenewedEvidence &&
                evidence.ShortTermScore >= 0.60 &&
                evidence.LongTermScore >= 0.65)
            {
                return new FactorLifecycleDecision(
                    FactorState.Shadow,
                    false,
                    true,
                    0.25,
                    "Renewed evidence permits Shadow review; risk expansion still requires promotion evidence.");
            }
            return new FactorLifecycleDecision(
                currentState,
                false,
                true,
                RiskMultiplier(currentState),
                "Risk expansion remains blocked pending renewed evidence.");
        }

        return new FactorLifecycleDecision(
            currentState,
            false,
            currentState != FactorState.Active,
            RiskMultiplier(currentState),
            "Evidence is stable; no lifecycle contraction is required.");
    }

    private static double RiskMultiplier(FactorState state)
    {
        return state switch
        {
            FactorState.Active => 1,
            FactorState.Shadow => 0,
            FactorState.Candidate => 0,
            FactorState.Reduced => 0.50,
            FactorState.Probation => 0.15,
            FactorState.Retired => 0,
            FactorState.Quarantined => 0,
            _ => 0
        };
    }
}
